File: p104-maximum depth of binary tree.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

TreeNode* BuildTree(vector<int> &tree){
	if(tree.empty())	return NULL;
	queue<TreeNode*> que;
	TreeNode *root = new TreeNode(tree[0]);
	que.push(root);
	int i=1;
	while(i < tree.size()){
		if(tree[i] != -1){
			que.front()->left = new TreeNode(tree[i]);
			que.push(que.front()->left);
		}
		if(i+1 < tree.size() && tree[i+1] != -1){
			que.front()->right = new TreeNode(tree[i+1]);
			que.push(que.front()->right);
		}
		que.pop();
		i += 2;
	}
	return root;
}

void DFS(TreeNode *node, int lev, int &maxi){
	if(!node)	return;
	if(lev > maxi)	maxi = lev;
	DFS(node->left, lev+1, maxi);
	DFS(node->right, lev+1, maxi);
}

int Solution(TreeNode *root){
	int maxi = 0;
	DFS(root, 1, maxi);
	return maxi;
}

int main(){
	int T;
	cin >> T;
	vector<int> tree(T);
	for(int i=0; i<T; i++)
		cin >> tree[i];
	TreeNode *root = BuildTree(tree);
	cout << Solution(root) << endl;
	return 0;
}

File: Python/p104-maximum depth of binary tree.py
from collections import deque


class TreeNode:
    def __init__(self, val=-1):
        self.val = val
        self.left = None
        self.right = None


def BuildTree(tree):
    if len(tree) == 0:
        return None
    que, root = deque(), TreeNode(tree[0])
    que.append(root)
    i = 1
    while i < len(tree):
        node = que.popleft()
        if tree[i] != -1:
            node.left = TreeNode(tree[i])
            que.append(node.left)
        if i+1 < len(tree) and tree[i+1] != -1:
            node.right = TreeNode(tree[i+1])
            que.append(node.right)
        i += 2
    return root

class Solution:
    def solve(self, root):
        ans = 0
        if root is None:
            return ans
        que = deque([root])
        while len(que) > 0:
            n = len(que)
            ans += 1
            for i in range(n):
                node = que.popleft()
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
        return ans


tree = [int(v) for v in input().split()]
solver = Solution()
print(solver.solve(BuildTree(tree)))
